### PR TITLE
Admin command to manage project network

### PR DESCRIPTION
**Pull request by pravisankar**
- Join project networks:
    * oadm pod-network join-projects --to=[project-name] [list of projects] | --selector=[project-labels]
- Unisolate project networks:
    * oadm pod-network unisolate-projects [list of projects] | --selector=[project-labels]
- Isolate project networks: 
    * TBD (some preliminary work done)

- Persist Net namespaces with VNID=0
- Release VNID only if no namespace is using it
- openshift network plugins can handle non pod infra container
- openshift-ovs-multitenant/openshift-ovs-subnet cleanup

### CODE REVIEW

**Comment by pravisankar** [comment]
@danwinship @dcbw @marun  PTAL

File: pkg/cmd/admin/network/isolate_project.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	IsolateProjectNetworkCommandName = "isolate-projects"
+
+	isolateProjectNetworkLong = `
+Isolate project network
+
+Allows project to isolate their network from other projects.`
```
Comment (danwinship) [inline pkg/cmd/admin/network/isolate_project.go]: Needs to clarify that this only works with the multitenant plugin. ie, it only undoes "unisolate-projects".

(Inconsistency between "project" and "projects"; filename and most variables use the singular, but the actual option uses plural.)

File: pkg/cmd/admin/network/join_project.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	JoinProjectNetworkCommandName = "join-projects"
+
+	joinProjectNetworkLong = `
+Join project network
+
+Allow projects to join existing project network.`
```
Comment (danwinship) [inline pkg/cmd/admin/network/join_project.go]: again probably need to mention multitenant plugin

File: pkg/cmd/admin/network/join_project.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	projects, err := j.Options.GetProjects()
+	if err != nil {
+		return err
+	}
+
+	errList := []error{}
+	for _, project := range projects {
+		if netID == ovssubnet.VNID0 {
+			err = j.Options.DeleteNetNamespace(project.ObjectMeta.Name)
```
Comment (danwinship) [inline pkg/cmd/admin/network/join_project.go]: won't that cause ovssubnet/common.go to just allocate a new VNID for it on next startup?

File: pkg/cmd/admin/network/unisolate_project.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/spf13/cobra"
+
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	kerrors "k8s.io/kubernetes/pkg/util/errors"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	UnIsolateProjectNetworkCommandName = "unisolate-projects"
```
Comment (danwinship) [inline pkg/cmd/admin/network/unisolate_project.go]: I still don't like "unisolate" as the name for this operation, since I think it only implies "everyone can access this project", not "this project can access everyone". Something with "admin" in the name would be better IMHO

File: pkg/ovssubnet/common.go [hunk 5]
```diff
@@ -21,6 +21,8 @@ import (
 const (
 	// Maximum VXLAN Network Identifier as per RFC#7348
 	MaxVNID = ((1 << 24) - 1)
+	// VNID for the admin namespaces
+	VNID0 = uint(0)
```
Comment (danwinship) [inline pkg/ovssubnet/common.go]: "AdminVNID"? If we think it might ever be something other than 0, then "VNID0" is a bad name, and if we don't think it might ever be non-0, then there's no reason to not just hardcode 0...

File: pkg/ovssubnet/controller/kube/bin/openshift-ovs-subnet [hunk 6]
```diff
@@ -34,13 +34,16 @@ Setup() {
...
     cluster_subnet=${OPENSHIFT_CLUSTER_SUBNET}
 
-    pid=$(docker inspect --format "{{.State.Pid}}" ${net_container})
     network_mode=$(docker inspect --format "{{.HostConfig.NetworkMode}}" ${net_container})
     if [ "${network_mode}" == "host" ]; then
       # quit, nothing for the SDN here
       exit 0
+    elif [[ "${network_mode}" =~ container:.* ]]; then
+      # Get pod infra container
+      net_container=$(echo ${network_mode} | cut -d ":" -f 2)
```
Comment (danwinship) [inline pkg/ovssubnet/controller/kube/bin/openshift-ovs-subnet]: why is this needed? shouldn't all containers in the pod show the same IP address?

**Comment by dcbw** [comment]
Instead of "unisolate" could we do "split" or "leave"?

**Comment by danwinship** [comment]
That's not what unisolate does. Unisolate means "set VNID to 0".
(If you want to split two previously-merged projects A and B, the only way to do that in this UI is to create a new project C, join B to C, and then delete C.)

**Comment by dcbw** [comment]
Yeah, looking at the patches I see there's also an unimplemented "isolate" command.  Which is what I was missing.

Still, seems odd that there isn't a corresponding opposite to Join, like Leave or Split...  If Isolate will eventually create a new VNID for a project/namespace, and Unisolate moves it back into the admin/VNID0 namespace, would the corresponding Leave/Split instead just be calling Isolate on the currently-joined project?

**Comment by danwinship** [comment]
what about:

join (change VNID of PROJECTS[1:] to VNID of PROJECTS[0])
`oadm pod-network join PROJECTS...`
split/isolate: (assign a distinct VNID to each project in PROJECTS)
`oadm pod-network isolate PROJECTS...`
unisolate: (assign VNID 0 to each project in PROJECTS)
`oadm pod-network admin PROJECTS...`

File: pkg/ovssubnet/controller/kube/bin/openshift-ovs-subnet (hunk 6 above)
Comment (pravisankar) [inline pkg/ovssubnet/controller/kube/bin/openshift-ovs-subnet]: If we do nsenter into the container, then yes, all the containers will show the same IP address but docker inspect on non infra pod container won't show the IP addr instead it's NetworkMode points to the pod infra container. I tried to avoid nsenter + parsing in favor of docker inspect on the pod infra container.

File: pkg/cmd/admin/network/join_project.go (hunk 3 above)
Comment (pravisankar) [inline pkg/cmd/admin/network/join_project.go]: Good catch, do you foresee any issues if we store VNID=0 in the NetNamespace registry? I don't understand why we did not persist namespaces with VNID=0 in the first place.

File: pkg/cmd/admin/network/isolate_project.go (hunk 1 above)
Comment (pravisankar) [inline pkg/cmd/admin/network/isolate_project.go]: I'll fix this one. I need to check how we can get network plugin name in pod-network command, that way we can provide useful error message when this cmd is used for flat sdn.

**Comment by dcbw** [comment]
> what about:
> 
> join (change VNID of PROJECTS[1:] to VNID of PROJECTS[0])
> oadm pod-network join PROJECTS...
> split/isolate: (assign a distinct VNID to each project in PROJECTS)
> oadm pod-network isolate PROJECTS...
> unisolate: (assign VNID 0 to each project in PROJECTS)
> oadm pod-network admin PROJECTS...

That looks a bit clearer to me.  What do you think @pravisankar ?

File: pkg/cmd/admin/network/join_project.go (hunk 3 above)
Comment (danwinship) [inline pkg/cmd/admin/network/join_project.go]: I'm not sure why we didn't either.

**Comment by pravisankar** [comment]
We had some discussion on this before in the email chain, let me recap:
We are creating overlay network for the pods and 'pod-network' was chosen to avoid any confusion with node network.
Eventually 'pod-network' command will support many sub commands once we have granular network policy. 
'-projects' suffix was added to indicate these sub-commands are operated on the projects.
I think, join-projects/isolate-projects provides clear semantics. For join projects, --to option is explicitly used to eliminate any confusion on which project to join, first one or the last one and this pattern is used in other openshift cmds.
When we have more than 2 projects sharing the pod network, split operation might be confusing. splitting from given project or existing pod network? I think, 'isolate' word is less confusing!
For VNID 0 projects:  'unisolate' was chosen to imply 'no isolation' and the help message clears any confusion. 'admin'/'global' are generic, they have other connotations. 'admin network' also needs explanation and this is completely different from the admin we use in ACLs. We need better alternatives if we are not happy with 'unisolate'.

**Comment by pravisankar** [comment]
Updated, PTAL

File: pkg/cmd/admin/network/isolate_projects.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	IsolateProjectsNetworkCommandName = "isolate-projects"
+
+	isolateProjectsNetworkLong = `
+Isolate project network
+
+Allows projects to isolate their network from other projects in case of %[1]s network plugin.`
```
Comment (danwinship) [inline pkg/cmd/admin/network/isolate_projects.go]: I think "Allows projects to isolate their network from other projects (when using the %[1]s network plugin)." would be clearer.

**Comment by danwinship** [comment]
> When we have more than 2 projects sharing the pod network, split operation might be confusing. splitting from given project or existing pod network? I think, 'isolate' word is less confusing!

I think in the case where you pass more than one project, it's not obvious whether that means they all become isolated from each other (each gets its own new VNID), or they all become isolated from everyone else (they all get the same new VNID). The example suggests it's the latter, but I would have guessed the former, since having them share the same VNID means "isolate" also does a "join".

(At any rate, that code isn't hooked up yet, so it doesn't matter that much today.)

> For VNID 0 projects: 'unisolate' was chosen to imply 'no isolation' and the help message clears any confusion.

I think that violates the ordinary meaning of "isolation" though. If project A is described as being "isolated", and project B as "not isolated", no one would expect that to mean that B has full access to A. Because if B has full access to A, then A isn't really isolated. That's why I was suggesting something like "admin"; then we can say that A really is isolated, it's just that B has superpowers and can bypass the isolation. (Although I guess "admin" has the reverse problem; it implies "I can access everyone else", but doesn't imply "everyone else can access me"...)

Let's get the code in at least. Presumably we can still potentially rename it post-dev-cut

**Comment by dcbw** [comment]
Agreed, I had no real problems with the code, just the UX.

**Comment by pravisankar** [comment]
On Mon, Oct 12, 2015 at 9:07 AM, Dan Winship <notifications@github.com>
wrote:

> When we have more than 2 projects sharing the pod network, split operation
> might be confusing. splitting from given project or existing pod network? I
> think, 'isolate' word is less confusing!
>
> I think in the case where you pass more than one project, it's not obvious
> whether that means they all become isolated from each other (each gets its
> own new VNID), or they all become isolated from everyone else (they all get
> the same new VNID). The example suggests it's the latter, but I would have
> guessed the former, since having them share the same VNID means "isolate"
> also does a "join".
>
> (At any rate, that code isn't hooked up yet, so it doesn't matter that
> much today.)
>
> For VNID 0 projects: 'unisolate' was chosen to imply 'no isolation' and
> the help message clears any confusion.
>
> I think that violates the ordinary meaning of "isolation" though. If
> project A is described as being "isolated", and project B as "not
> isolated", no one would expect that to mean that B has full access to A.
> Because if B has full access to A, then A isn't really isolated. That's why
> I was suggesting something like "admin"; then we can say that A really is
> isolated, it's just that B has superpowers and can bypass the isolation.
> (Although I guess "admin" has the reverse problem; it implies "I can access
> everyone else", but doesn't imply "everyone else can access me"...)
>
Instead of admin, may be we can use 'global' (making the project global
within the cluster). Something like 'make-global-projects'?

> Let's get the code in at least. Presumably we can still potentially rename
> it post-dev-cut
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/openshift-sdn/pull/176#issuecomment-147447697>
> .
>



-- 
-Ravi
